File: app-presences/src/main/java/fr/openent/presences/service/DisciplineService.java
package fr.openent.presences.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;


public interface DisciplineService {

    /**
     * get disciplines
     *
     * @param structureId Structure Identifier
     * @param handler     Function handler returning data
     */
    void get(String structureId, Handler<Either<String, JsonArray>> handler);

    /**
     * create discipline
     *
     * @param disciplineBody disciplineBody fetched
     * @param handler        Function handler returning data
     */
    void create(JsonObject disciplineBody, Handler<Either<String, JsonObject>> handler);

    /**
     * put discipline
     *
     * @param disciplineBody disciplineBody fetched
     * @param handler        Function handler returning data
     */
    void put(JsonObject disciplineBody, Handler<Either<String, JsonObject>> handler);

    /**
     * delete discipline
     *
     * @param disciplineId discipline identifier
     * @param handler      Function handler returning data
     */
    void delete(Integer disciplineId, Handler<Either<String, JsonObject>> handler);
}

File: presences/src/main/resources/public/ts/services/DisciplineService.ts
import {ng} from 'entcore'
import { http, HttpResponse } from 'entcore-toolkit';
import {Discipline, DisciplineRequest} from "../models";

export interface DisciplineService {
    get(structureId: string): Promise<Discipline[]>;

    create(disciplineBody: DisciplineRequest): Promise<HttpResponse>;

    update(disciplineBody: DisciplineRequest): Promise<HttpResponse>;

    delete(disciplineId: number): Promise<HttpResponse>;
}

export const disciplineService: DisciplineService = {
    get: async (structureId: string): Promise<Discipline[]> => {
        try {
            const {data} = await http.get(`/presences/disciplines?structureId=${structureId}`);
            return data;
        } catch (err) {
            throw err;
        }
    },

    create: async (disciplineBody: DisciplineRequest): Promise<HttpResponse> => {
        return http.post(`/presences/discipline`, disciplineBody);
    },

    update: async (disciplineBody: DisciplineRequest): Promise<HttpResponse> => {
        return http.put(`/presences/discipline`, disciplineBody);
    },

    delete: async (disciplineId: number): Promise<HttpResponse> => {
        return http.delete(`/presences/discipline?id=${disciplineId}`);
    },
};

export const DisciplineService = ng.service('DisciplineService', (): DisciplineService => disciplineService);
